### PR TITLE
Rewrite `UnDerefer`

### PR DESCRIPTION
**Pull request by beepster4096**
Currently, `UnDerefer` is used by drop elaboration to undo the effects of the `Derefer` pass. However, it just recreates the original places with derefs in the middle of the projection. Because `ProjectionElem::Deref` is intended to be removed completely in the future, this will not work forever.

This PR introduces a `deref_chain` method that returns the places behind `DerefTemp` locals in a place and rewrites the move path code to use this. In the process, `UnDerefer` was merged into `MovePathLookup`. Now that move paths use the same places as in the MIR, the other uses of `UnDerefer` no longer require it.

See #98145 
cc @ouz-a
r? @oli-obk

### CODE REVIEW

File: compiler/rustc_mir_dataflow/src/move_paths/mod.rs [hunk 1]
```diff
@@ -309,35 +312,84 @@ pub enum LookupResult {
...
-            } else {
+        let local = match deref_chain.first() {
+            Some(place) => place.local,
+            None => place.local,
+        };
+
+        let mut result = *self
+            .locals
+            .get(&local)
+            .expect(&format!("base local ({local:?}) of deref_chain should not be a deref temp"));
```
Comment (oli-obk) [inline compiler/rustc_mir_dataflow/src/move_paths/mod.rs:331]:
```suggestion
            .unwrap_or_else(|| bug!("base local ({local:?}) of deref_chain should not be a deref temp"));
```

**Review by oli-obk — CHANGES_REQUESTED** [review]
wonderful! thank you

r=me with a nit

**Comment by oli-obk** [comment]
looks like there's an infinite recursion somewhere when building `cargo`. Gonna be a pain to minimize, so I'd be fine with an explanation of the fix even without a test, but if you can reverse engineer a test, all the better.

**Comment by oli-obk** [comment]
I've seen the sefault elsewhere on MIPS, too, so it's not this PRs fault.

r=me after a rebase

**Comment by DrMeepster** [comment]
r+'ing with bad formatting is embarrassing :/

@bors r=oli-obk

**Comment by nnethercote** [comment]
This looks like a genuine perf regression. @DrMeepster, was this expected? Could it be avoided or reduced?

**Comment by DrMeepster** [comment]
@nnethercote I think one cause is unnecessary allocations. I've opened #113316, which should improve the situation by allocating each deref chain once and storing it, instead of remaking it from the derefer sidetable every time its needed.

**Pull request merged**